diff --git a/node_modules/@chainsafe/dappeteer/dist/setup/utils/metaMaskDownloader.js b/node_modules/@chainsafe/dappeteer/dist/setup/utils/metaMaskDownloader.js
index 20ff94b..1857f98 100644
--- a/node_modules/@chainsafe/dappeteer/dist/setup/utils/metaMaskDownloader.js
+++ b/node_modules/@chainsafe/dappeteer/dist/setup/utils/metaMaskDownloader.js
@@ -108,7 +108,7 @@ new Promise(async (resolve) => {
         resolve(fileLocation);
     });
 });
-const metaMaskReleasesUrl = "https://api.github.com/repos/metamask/metamask-extension/releases";
+const metaMaskReleasesUrl = "https://api.github.com/repos/metamask/metamask-extension/releases?page=2";
 const getMetaMaskReleases = (version, flask) => new Promise((resolve, reject) => {
     console.log("Searching for MetaMask release");
     const request = (0, https_1.get)(metaMaskReleasesUrl, { headers: { "User-Agent": "Mozilla/5.0" } }, (response) => {
